chatmeleon-server/src/controllers/message-controller.ts

```diff
@@ -1,114 +1,252 @@
-
-  if (!(await hasConversation(req.auth!.userId, conversationId))) {
-    return res
-      .status(403)
-      .type("text/plain")
-      .send(
-        "Forbidden - User not allowed to retrieve the messages of the conversation."
-      );
-  }
-
-  const allMessages = await prisma.message.findMany({
-    where: {
-      conversation: {
-        id: conversationId,
-      },
-      id: {
-        gt: cursor,
-      },
-    },
-    take: pageSize,
-  });
+  const { userId } = req.auth!; // Get authenticated user's ID
+  const conversationId = req.params.conversationId;
+  const pageSize = parseInt(req.query.pageSize as string, 10) || 10; // Get desired page size from query parameters
+  const cursor = req.query.cursor as string | undefined; // Get optional cursor for pagination from query parameters
 
-  const lastMessage = allMessages[allMessages.length - 1];
-  const nextCursor = lastMessage ? lastMessage.id : null;
+  prisma
+    .$transaction(async (tx) => {
+      try {
+        // Validate conversation existence
+        await tx.conversation.findFirstOrThrow({
+          where: {
+            id: conversationId,
+            userIds: { has: userId },
+          },
+        });
 
-  res.json({
-    allMessages,
-    nextCursor,
-  });
+        const allMessages = await tx.message.findMany({
+          where: {
+            conversation: {
+              id: conversationId, // Filter messages for the conversation
+            },
+            id: {
+              gt: cursor, // Retrieve conversations after the provided cursor (if any)
+            },
+          },
+          orderBy: {
+            createdAt: "asc", // Order messages by creation date in ascending order
+          },
+          take: pageSize, // Limit results to the specified page size
+        });
+
+        // Extract the last messages's cursor for pagination.
+        const lastMessage = allMessages[allMessages.length - 1];
+        const nextCursor = lastMessage ? lastMessage.id : null;
+
+        return res.json({
+          allMessages, // Send the fetched messages in the response
+          nextCursor, // Include the cursor for further pagination
+        });
+      } catch (error) {
+        // Rethrow the error to trigger the transaction rollback
+        throw error;
+      }
+    })
+    .catch((error) => {
+      // Log the error and send a 403 Unauthorized response
+      console.error("Transaction failed:", error);
+      res
+        .status(403)
+        .type("text/plain")
+        .send("Unauthorized: User is unauthorized");
+    });
 };
 
-const getAllMessagesByUserId = async (req: Request, res: Response) => {
-  const userId = req.query.userId as string;
-  const pageSize = parseInt(req.query.pageSize as string, 10) || 10;
-  const cursor = req.query.cursor as string | undefined;
+// Fetch messages for the authorized user with pagination
+const getMessagesByUserIdWithPagination = async (
+  req: Request,
+  res: Response
+) => {
```

Comment:
Why do we need to get messages of a user, possibly sent in different conversations?

Reply:
Yes you are correct